storage/innobase/buf/buf0flu.cc

```diff
@@ -1885,6 +1915,35 @@ inline void log_t::write_checkpoint(lsn_t end_lsn) noexcept
     resize_lsn.store(0, std::memory_order_relaxed);
   }
 
+#if defined HAVE_INNODB_MMAP && defined FALLOC_FL_ZERO_RANGE
+  if (is_mmap())
+  {
+    const size_t ps{my_system_page_size};
+    ut_ad(buf_free == calc_lsn_offset(get_lsn()));
+    size_t offset{buf_free & ~(ps - 1)};
+    const size_t start_offset{calc_lsn_offset(checkpoint_lsn) & ~(ps - 1)};
+    if (offset == start_offset);
+    else if (offset < start_offset)
+    {
+      madvise(buf + offset, start_offset - offset, MADV_DONTNEED);
+      fallocate(log.m_file, FALLOC_FL_ZERO_RANGE, offset,
+                start_offset - offset);
```

Comment:
Based on my tests on a HDD (see [MDEV-34062](https://jira.mariadb.org/browse/MDEV-34062)), it seems that the `fallocate`, which aims to prevent a read-on-write problem for a subsequent round of writing the circular `ib_logfile0`, is making things worse.

The `MADV_DONTNEED`, which is available on POSIX, seems to help in the `innodb_log_file_buffering=ON` (no `O_DIRECT`) case when the entire `ib_logfile0` apparently fits in the Linux file system or block cache. So, I would remove the references to `FALLOC_FL_ZERO_RANGE` and enable this code on all POSIX.

I agree that for many setups, enabling memory-mapped log writes may be a bad idea because page faults during writes would lead to reads of the log file. I think that we should offer this option nevertheless. Possibly, `madvise(MADV_SEQUENTIAL)` could help, but I can imagine that it could also request the entire `ib_logfile0` to be read into cache even when it is not needed.

Reply:
3d2ec55fd92078b4023ec39af50fcd30aaf8667c removes the references to `FALLOC_FL_ZERO_RANGE`.

In my tests, the impact of this read-before-write does not seem too bad. Besides, the combination of memory-mapped writes in `mariadbd` and memory-mapped reads in `mariadb-backup --backup` can give a significant performance improvement, depending on the workload.

Reply:
@vaintroub says that the impact of read-before-write is bad on Microsoft Windows; 043745bd8df8954b8d449313dc4872d98c3ee4b6 includes some tweaks from his testing. So, we might want to limit the scope to reading the log, in `mariadb-backup` only.

Reply:
ea8749820ddfc3f8f42c9477b3e95e959fec8a5d addresses the issue on Microsoft Windows by not enabling memory-mapped log access in `mariadb-backup` by default. On other 64-bit platforms than Windows, it will be enabled by default. On Linux, if one specified `MAP_POPULATE`, there should be synchronous read-ahead. But we of course do not do that; we prefer on-demand paging.

I think that `innodb_log_file_mmap=ON` can make sense on any platform when the `ib_logfile0` is relatively small (users prefer fast recovery times, do not care about write performance, or there are not too many writes).